File: Data Structures & Algorithms/depth-of-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode], i: int = 1) -> int:
        if not root:
            return 0
            
        maxH = 1

        stack = [root]
        heights = [1]

        while stack:
            node = stack.pop()
            h = heights.pop()
            maxH = max(h, maxH)

            if node.left:
                stack.append(node.left)
                heights.append(h+1)
            
            if node.right:
                stack.append(node.right)
                heights.append(h+1)

        return maxH
        
        

File: Data Structures & Algorithms/validate-parentheses/submission-2.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> stack;
        for (char c : s) {
            if (c == '{' || c == '[' || c == '(') {
                stack.push(c);
            } else if (stack.empty()) {
                return false;
            } else if (c == '}' && stack.top() == '{') {
                stack.pop();
                continue;
            } else if (c == ')' && stack.top() == '(') {
                stack.pop();
                continue;
            } else if (c == ']' && stack.top() == '[') {
                stack.pop();
                continue;
            } else {
                return false;
            }
        }
        if (stack.empty())
            return true;
        return false;
    }
    
};
